# Movie Recommendation System using Machine Learning 

### Import Dependencies

In [75]:
import numpy as np 
import pandas as pd
import difflib 
from sklearn.feature_extraction.text import TfidfVectorizer # used to convert text into numeric value
from sklearn.metrics.pairwise import cosine_similarity

### data Collection and Data Pre-Processing

In [76]:
# loading the dataset from csv to pandas dataframe

movies_data = pd.read_csv('dataset/movies.csv')

In [77]:
# print first 5 rows of the data 

movies_data.head()

index     budget                                    genres  \
0      0  237000000  Action Adventure Fantasy Science Fiction   
1      1  300000000                  Adventure Fantasy Action   
2      2  245000000                    Action Adventure Crime   
3      3  250000000               Action Crime Drama Thriller   
4      4  260000000          Action Adventure Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  culture clash future space war space colony so...                en   
1  ocean drug abuse exotic island east india trad...                en   
2         spy based on novel secret agent sequel mi6                en   
3  dc comics crime fighter terrorist secret ident...                en   
4  based on novel mars medallion space travel pri...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ... runtime  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   162.0   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   169.0   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   148.0   
3  Following the death of District Attorney Harve...  112.312950  ...   165.0   
4  John Carter is a war-weary, former military ca...   43.926995  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
0  Sam Worthington Zoe Saldana Sigourney Weaver S...   
1  Johnny Depp Orlando Bloom Keira Knightley Stel...   
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...   
3  Christian Bale Michael Caine Gary Oldman Anne ...   
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...   

                                                crew           director  
0  [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...      James Cameron  
1  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski  
2  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes  
3  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan  
4  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton  

[5 rows x 24 columns]

In [78]:
# printing number of rows and columns in the dataframe 
movies_data.shape


(4803, 24)

In [79]:
# selecting the relevent features for recommendation 

selected_features= ['genres', 'keywords', 'tagline', 'cast', 'director' ]

print(selected_features)

['genres', 'keywords', 'tagline', 'cast', 'director']


In [80]:
# replacing null values with null string

for feature in selected_features:
    movies_data[feature] = movies_data[feature].fillna('')
    

In [81]:
# combining all the 5 selected features

combined_features = movies_data['genres'] + ' ' + movies_data['keywords'] + ' ' + movies_data['tagline'] + ' ' + movies_data['cast'] + ' ' + movies_data['director']

print(combined_features) 

0       Action Adventure Fantasy Science Fiction cultu...
1       Adventure Fantasy Action ocean drug abuse exot...
2       Action Adventure Crime spy based on novel secr...
3       Action Crime Drama Thriller dc comics crime fi...
4       Action Adventure Science Fiction based on nove...
                              ...                        
4798    Action Crime Thriller united states\u2013mexic...
4799    Comedy Romance  A newlywed couple's honeymoon ...
4800    Comedy Drama Romance TV Movie date love at fir...
4801      A New Yorker in Shanghai Daniel Henney Eliza...
4802    Documentary obsession camcorder crush dream gi...
Length: 4803, dtype: object


In [82]:
# converting the text data to feature vectors

vectorizer = TfidfVectorizer()

In [83]:
feature_vectors = vectorizer.fit_transform(combined_features)

In [84]:
print(feature_vectors)

  (0, 2432)	0.17272411194153
  (0, 7755)	0.11280357148547558
  (0, 13024)	0.19423620601088706
  (0, 10229)	0.160586854000953
  (0, 8756)	0.22709015857011813
  (0, 14608)	0.15150672398763912
  (0, 16668)	0.1984326396510037
  (0, 14064)	0.2059609041508414
  (0, 13319)	0.21774705394124838
  (0, 17290)	0.20197912553916564
  (0, 17007)	0.23643326319898794
  (0, 13349)	0.15021264094167083
  (0, 11503)	0.2721131005698365
  (0, 11192)	0.09049319826481456
  (0, 16998)	0.12821263228505786
  (0, 15261)	0.07095833561276564
  (0, 4945)	0.24025852494110755
  (0, 14271)	0.21392179219912874
  (0, 3225)	0.24960162956997733
  (0, 16587)	0.12549432354918996
  (0, 14378)	0.3396275221095982
  (0, 5836)	0.1646750903586285
  (0, 3065)	0.22208377802661422
  (0, 3678)	0.21392179219912874
  (0, 5437)	0.10364139873166359
  :	:
  (4801, 17266)	0.2886098184932947
  (4801, 4835)	0.24713765026963996
  (4801, 403)	0.17727585190343226
  (4801, 6935)	0.2886098184932947
  (4801, 11663)	0.21557500762727902
  (4801, 1672)

### Cosine Similarity 

#### Getting the similarity scores using cosine similarity

In [85]:
similarity = cosine_similarity(feature_vectors)

In [86]:
print(similarity)

[[1.         0.07219487 0.037733   ... 0.         0.         0.        ]
 [0.07219487 1.         0.03281499 ... 0.03575545 0.         0.        ]
 [0.037733   0.03281499 1.         ... 0.         0.05389661 0.        ]
 ...
 [0.         0.03575545 0.         ... 1.         0.         0.02651502]
 [0.         0.         0.05389661 ... 0.         1.         0.        ]
 [0.         0.         0.         ... 0.02651502 0.         1.        ]]


In [87]:
print(similarity.shape)

(4803, 4803)


(4803, 4803); first value is index ;  second value is similarity score;

In [88]:
# getting the movie name from the user 

movie_name = input(' Enter your favourite Movie Name : ')

print(movie_name)

iron man


In [89]:
# creating a list with all the movie titles in the dataset

list_of_all_titles = movies_data['title'].tolist()

print(list_of_all_titles)


['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre', 'The Dark Knight Rises', 'John Carter', 'Spider-Man 3', 'Tangled', 'Avengers: Age of Ultron', 'Harry Potter and the Half-Blood Prince', 'Batman v Superman: Dawn of Justice', 'Superman Returns', 'Quantum of Solace', "Pirates of the Caribbean: Dead Man's Chest", 'The Lone Ranger', 'Man of Steel', 'The Chronicles of Narnia: Prince Caspian', 'The Avengers', 'Pirates of the Caribbean: On Stranger Tides', 'Men in Black 3', 'The Hobbit: The Battle of the Five Armies', 'The Amazing Spider-Man', 'Robin Hood', 'The Hobbit: The Desolation of Smaug', 'The Golden Compass', 'King Kong', 'Titanic', 'Captain America: Civil War', 'Battleship', 'Jurassic World', 'Skyfall', 'Spider-Man 2', 'Iron Man 3', 'Alice in Wonderland', 'X-Men: The Last Stand', 'Monsters University', 'Transformers: Revenge of the Fallen', 'Transformers: Age of Extinction', 'Oz: The Great and Powerful', 'The Amazing Spider-Man 2', 'TRON: Legacy', 'Cars 2', 'Green Lant

In [90]:
# finding the close match for the movie name given by user

find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)

print(find_close_match)

['Iron Man', 'Iron Man 3', 'Iron Man 2']


In [91]:
close_match = find_close_match[0]
print(close_match)

Iron Man


In [92]:
# find the index of the movie with title

index_of_the_movie = movies_data[movies_data.title == close_match]['index'].values[0]

print(index_of_the_movie)


68


### Getting list of similar movies 

In [93]:
similarity_score = list(enumerate(similarity[index_of_the_movie]))

print(similarity_score)

len(similarity_score)

[(0, 0.03357074878067544), (1, 0.05464482792361339), (2, 0.013735500604224321), (3, 0.006468756104392058), (4, 0.03268943310073386), (5, 0.01390725668575547), (6, 0.07692837576335505), (7, 0.23944423963486405), (8, 0.007882387851851008), (9, 0.07599206098164224), (10, 0.07536074882460436), (11, 0.011926069211745289), (12, 0.013707618139948929), (13, 0.012376074925089967), (14, 0.09657127116284184), (15, 0.007286271383816741), (16, 0.22704403782296806), (17, 0.013112928084103853), (18, 0.04140526820609593), (19, 0.07883282546834251), (20, 0.07981173664799915), (21, 0.011266873271064947), (22, 0.006892575895462362), (23, 0.006599097891242657), (24, 0.012665208122549735), (25, 0.0), (26, 0.21566241096831154), (27, 0.03058128209382663), (28, 0.06107440221966537), (29, 0.014046184258938894), (30, 0.0807734659476981), (31, 0.31467052449477506), (32, 0.028782099134267002), (33, 0.13089810941050173), (34, 0.0), (35, 0.035350090674865595), (36, 0.03185325269937554), (37, 0.008024326882532318), 

4803

In [94]:
# sorting the movies based on similarity score

sorted_similar_movies = sorted(similarity_score, key = lambda x:x[1], reverse = True )

print(sorted_similar_movies)


[(68, 0.9999999999999996), (79, 0.4089043399800596), (31, 0.31467052449477506), (7, 0.23944423963486405), (16, 0.22704403782296806), (26, 0.21566241096831154), (85, 0.20615862984665329), (182, 0.19573956139611606), (511, 0.16702973947860686), (3623, 0.16092460881355858), (64, 0.15299924139445137), (203, 0.14818667948665118), (174, 0.1471993120942043), (4401, 0.14505971470107845), (101, 0.14401677581826292), (46, 0.14216268867232235), (169, 0.1380947013224906), (1740, 0.13624382641690763), (94, 0.1361681957902901), (788, 0.13305895074229218), (126, 0.13263982780511066), (131, 0.13137698586006535), (33, 0.13089810941050173), (2487, 0.12309731939910507), (783, 0.12162995562040374), (138, 0.11846458075866885), (2442, 0.11725512335483318), (661, 0.11719294096248459), (607, 0.11387063493435635), (38, 0.1126182690487113), (2651, 0.11218787873732047), (353, 0.11168465127044278), (122, 0.10850296033661253), (1553, 0.10797822171513259), (1451, 0.107849394974707), (242, 0.1063033902232701), (618,

In [95]:
print(sorted_similar_movies[:6])

[(68, 0.9999999999999996), (79, 0.4089043399800596), (31, 0.31467052449477506), (7, 0.23944423963486405), (16, 0.22704403782296806), (26, 0.21566241096831154)]


In [96]:
# print the name of the similar movies based on the index

print('Movies suggested for you : \n')

i = 1

for movie in sorted_similar_movies:
    index = movie[0]
    title_from_index = movies_data[movies_data.index == index]['title'].values[0]
    
    if (i < 21):
        print(i , '.', title_from_index)
        i += 1

Movies suggested for you : 

1 . Iron Man
2 . Iron Man 2
3 . Iron Man 3
4 . Avengers: Age of Ultron
5 . The Avengers
6 . Captain America: Civil War
7 . Captain America: The Winter Soldier
8 . Ant-Man
9 . X-Men
10 . Made
11 . X-Men: Apocalypse
12 . X2
13 . The Incredible Hulk
14 . The Helix... Loaded
15 . X-Men: First Class
16 . X-Men: Days of Future Past
17 . Captain America: The First Avenger
18 . Kick-Ass 2
19 . Guardians of the Galaxy
20 . Deadpool


## Movie Recommendation system

In [99]:
# getting the movie name from the user 

movie_name = input(' Enter your favourite Movie Name : ')

print(movie_name)

# creating a list with all the movie titles in the dataset

list_of_all_titles = movies_data['title'].tolist()

# finding the close match for the movie name given by user

find_close_match = difflib.get_close_matches(movie_name, list_of_all_titles)

close_match = find_close_match[0]

# find the index of the movie with title

index_of_the_movie = movies_data[movies_data.title == close_match]['index'].values[0]

similarity_score = list(enumerate(similarity[index_of_the_movie]))

# sorting the movies based on similarity score

sorted_similar_movies = sorted(similarity_score, key = lambda x:x[1], reverse = True )

print('Movies suggested for you : \n')

i = 1

for movie in sorted_similar_movies:
    index = movie[0]
    title_from_index = movies_data[movies_data.index == index]['title'].values[0]
    
    if (i < 21):
        print(i , '.', title_from_index)
        i += 1

avatar
Movies suggested for you : 

1 . Avatar
2 . Alien
3 . Aliens
4 . Guardians of the Galaxy
5 . Star Trek Beyond
6 . Star Trek Into Darkness
7 . Galaxy Quest
8 . Alien³
9 . Cargo
10 . Trekkies
11 . Gravity
12 . Moonraker
13 . Jason X
14 . Pocahontas
15 . Space Cowboys
16 . The Helix... Loaded
17 . Lockout
18 . Event Horizon
19 . Space Dogs
20 . Machete Kills
